haystack/database/faiss.py

```diff
@@ -0,0 +1,173 @@
+import logging
+from pathlib import Path
+from typing import Union, List, Optional
+
+import faiss
+import numpy as np
+from faiss.swigfaiss import IndexHNSWFlat
+
+from haystack.database.base import Document
+from haystack.database.sql import SQLDocumentStore
+from haystack.retriever.base import BaseRetriever
+
+logger = logging.getLogger(__name__)
+
+
+class FAISSDocumentStore(SQLDocumentStore):
+    """
+    Document store for very large scale embedding based dense retrievers like the DPR.
+
+    It implements the FAISS library(https://github.com/facebookresearch/faiss)
+    to perform similarity search on vectors.
+
+    The document text and meta-data(for filtering) is stored using the SQLDocumentStore, while
+    the vector embeddings are indexed in a FAISS Index.
+
+    """
+
+    def __init__(
+        self,
+        sql_url: str = "sqlite:///",
+        index_factory: str = "HNSW4",
+        index_buffer_size: int = 10_000,
+        vector_size: int = 768,
+    ):
+        """
+        :param sql_url: SQL connection URL for database. It defaults to local file based SQLite DB. For large scale
+                        deployment, Postgres is recommended.
+        :param index_factory: FAISS provides a function to build composite index based on comma separated list of
+                              components. The FAISS documentation has guidelines for choosing
+                              an index: https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index.
+                              More details on index_factory: https://github.com/facebookresearch/faiss/wiki/The-index-factory
+
+                              The default index here is fast with good accuracy at the cost of higher memory usage. It
+                              does not require any training.
+        :param index_buffer_size: When working with large dataset, the indexing process(FAISS + SQL) can be buffered in
+                                  smaller chunks to reduce memory footprint.
+        :param vector_size: the embedding vector size.
+        """
+        self.vector_size = vector_size
+        self.faiss_index: IndexHNSWFlat = self._create_new_index(index_factory=index_factory, vector_size=vector_size)
+        self.index_factory = index_factory
+
+        self.index_buffer_size = index_buffer_size
+        self.phi = 0
```

Comment:
lets rename this self.phi parameter to self.index_filled and make it boolean. To me it seems it is only used to not index into an existing index again.

And self.phi seems different from the actual "phi" value we add to self._get_hnsw_vectors(embeddings=embeddings, phi=phi)